### PR TITLE
perf(polys): speed up AlgebraicField.from_sympy

### PR DESCRIPTION
**Pull request by oscarbenjamin**
Reduce the number of calls to to_number_field by caching the results when converting from Expr to a number field.

#### Brief description of what is fixed or changed

Cache calls to `to_number_field` in `AlgebraicField.from_sympy` so that we don't have to solve the full isomorphism problem every time. This makes conversions faster e.g. with the PR:
```python
In [3]: M = randMatrix(100)*sqrt(2) + randMatrix(100)*sqrt(3) + randMatrix(100)*sqrt(5)

In [4]: K = QQ.algebraic_field(sqrt(2),sqrt(3),sqrt(5))

In [5]: %time ok = M.to_DM(domain=K)
CPU times: user 2.65 s, sys: 8.16 ms, total: 2.65 s
Wall time: 2.65 s
```
With master this would take a lot longer...

#### Release Notes

* polys
   * Converting expressions into a number field is made faster by caching the results of `to_number_field`.

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
> With master this would take a lot longer...

I estimate that it would take about an hour. So the change here makes it about 1000x faster for the case shown.

File: sympy/polys/domains/algebraicfield.py [hunk 1]
```diff
@@ -569,6 +562,59 @@ def galois_group(self, by_name=False, max_tries=30, randomize=False):
...
+    # point we have forgotten that conversion...
+
+    def isomorphism_converter(a):
+        """Convert a to K by solving the field membership problem."""
+        from sympy.polys.numberfields import to_number_field
+        try:
+            res = K(to_number_field(a, K.ext).native_coeffs())
+        except (NotAlgebraic, IsomorphismFailed):
+            raise CoercionFailed(
+                "%s is not a valid algebraic number in %s" % (a, K))
```
Comment (Skylion007) [inline sympy/polys/domains/algebraicfield.py:586]:
```suggestion
        except (NotAlgebraic, IsomorphismFailed) as exc:
            raise CoercionFailed(
                "%s is not a valid algebraic number in %s" % (a, K)) from exc
```

**Comment by oscarbenjamin** [comment]
This is ready for review.

File: sympy/polys/domains/algebraicfield.py [hunk 2]
```diff
@@ -569,6 +562,59 @@ def galois_group(self, by_name=False, max_tries=30, randomize=False):
...
+    # calls to it by reducing down to the basic algebraic elements and then
+    # cache the actual calls to to_number_field. E.g 2*sqrt(2) + 3*sqrt(3) will
+    # be reduced to sqrt(2) and sqrt(3) and then to_number_field will be called
+    # on those using the cached converter.
+    #
+    # XXX: Most of the time the calls to isomorphism_converter are unnecessary.
+    # They result from e.g. QQ.algebraic_field(sqrt(2), sqrt(3)) trying to
+    # convert sqrt(2) and sqrt(3) to the field. When primitive_element is
+    # called it computes the conversion for the original elements but by this
+    # point we have forgotten that conversion...
```
Comment (skieffer) [inline sympy/polys/domains/algebraicfield.py:577]: Do you suppose it would be worth caching the conversion?

The field's primitive element is computed here:

https://github.com/sympy/sympy/blob/ea1e3852851a9a84f124dea874307fa9ae11ab95/sympy/polys/domains/algebraicfield.py#L293-L302

with a call to `to_number_field()`, which is sort of a Swiss Army Knife function that ultimately uses `primitive_element()`. The field constructor could instead call `primitive_element()` directly, and could then opt to pass `ex=True`, in order to obtain the desired representations.

Currently `ex=False` is used, meaning the representations aren't just *forgotten* by now; they were never actually computed. I wonder if computing them is significantly slower, or if it would be worth obtaining them at construction time.

File: sympy/polys/domains/algebraicfield.py [hunk 3]
```diff
@@ -351,25 +351,18 @@ def to_alg_num(self, a):
...
         """Convert ``a`` of ``dtype`` to a SymPy object. """
         # Precompute a converter to be reused:
-        if not hasattr(self, '_converter'):
-            self._converter = _make_converter(self)
-
-        return self._converter(a)
+        conv = getattr(self, '_converter', None)
+        if conv is None:
+            conv = self._converter = _make_converter(self)
+        return conv(a)
```
Comment (skieffer) [inline sympy/polys/domains/algebraicfield.py:357]: This just looks like a rephrasing. Is it more efficient?

File: sympy/polys/domains/algebraicfield.py (hunk 2 above)
Comment (oscarbenjamin) [reply sympy/polys/domains/algebraicfield.py:577]: I'm not sure. The code in `primitive_element` is here:
https://github.com/sympy/sympy/blob/ea1e3852851a9a84f124dea874307fa9ae11ab95/sympy/polys/numberfields/subfield.py#L350-L388
I would have guessed that the expensive part is `_factor_list` and potentially `_choose_factor` (if it needs to numerically evaluate RootOf).

The extra pieces to generate the representation of the elements seem like they would be cheap in the grand scheme of things. It is difficult to measure though because so many things are cached so these timings may not be representative:
```python
In [1]: surds = lambda n: [sqrt(prime(i)) for i in range(1, n+1)]

In [2]: %time ok = primitive_element(surds(2))
CPU times: user 106 ms, sys: 0 ns, total: 106 ms
Wall time: 105 ms

In [3]: %time ok = primitive_element(surds(3))
CPU times: user 67.2 ms, sys: 3.94 ms, total: 71.1 ms
Wall time: 70 ms

In [4]: %time ok = primitive_element(surds(4))
CPU times: user 82.7 ms, sys: 3.6 ms, total: 86.3 ms
Wall time: 84.8 ms

In [5]: %time ok = primitive_element(surds(5))
CPU times: user 147 ms, sys: 0 ns, total: 147 ms
Wall time: 145 ms

In [6]: %time ok = primitive_element(surds(6))
CPU times: user 315 ms, sys: 132 μs, total: 315 ms
Wall time: 317 ms
```
Compared to:
```python
In [1]: surds = lambda n: [sqrt(prime(i)) for i in range(1, n+1)]

In [2]: %time ok = primitive_element(surds(2), ex=True)
CPU times: user 181 ms, sys: 11.5 ms, total: 193 ms
Wall time: 191 ms

In [3]: %time ok = primitive_element(surds(3), ex=True)
CPU times: user 53.4 ms, sys: 87 μs, total: 53.4 ms
Wall time: 52.2 ms

In [4]: %time ok = primitive_element(surds(4), ex=True)
CPU times: user 88.6 ms, sys: 73 μs, total: 88.7 ms
Wall time: 87.7 ms

In [5]: %time ok = primitive_element(surds(5), ex=True)
CPU times: user 141 ms, sys: 3.77 ms, total: 144 ms
Wall time: 144 ms

In [6]: %time ok = primitive_element(surds(6), ex=True)
CPU times: user 423 ms, sys: 4.26 ms, total: 427 ms
Wall time: 429 ms
```
In general in SymPy is there ever a situation where we call `primitive_element` but would not want `ex=True`?

It seems like it is never going to be sufficient just to have the primitive element if you don't also have a way to represent the components of the primitive element in the number field.

File: sympy/polys/domains/algebraicfield.py (hunk 3 above)
Comment (oscarbenjamin) [reply sympy/polys/domains/algebraicfield.py:357]: The `hasattr` call does an attribute lookup and so does `self._converter`. With `getattr` there is only one loop up. This is a very minor point and likely not significant compared to the cost of actually calling `conv` but in general there is a computational cost in an attribute lookup.

File: sympy/polys/domains/algebraicfield.py (hunk 2 above)
Comment (skieffer) [reply sympy/polys/domains/algebraicfield.py:577]: It's hard to say. The `ex=False` default was already in there when I refactored `polys.numberfields` a few years ago; i.e. it was not my design choice.

I agree though, it does seem reasonable that, if you're starting with multiple generators for a field, then you're going to want to continue to work with those generators, once you have the field.

For example, the fact that the `AlgebraicField` class stores the original generators under `self.orig_ext` attests to their continued importance.

But it's hard to guess whether the proposed change represents an overall performance penalty or boost for the average user.

Bottom line: I think this PR is fine as is. But if you wanted to try caching the generator representations at build time, I think I'd support that too (unless someone comes in with a great argument against it).

File: sympy/polys/domains/algebraicfield.py (hunk 3 above)
Comment (skieffer) [reply sympy/polys/domains/algebraicfield.py:357]: Sounds good.

File: sympy/polys/domains/algebraicfield.py (hunk 2 above)
Comment (oscarbenjamin) [reply sympy/polys/domains/algebraicfield.py:577]: Most of the time all the elements are converted to the domain at the same after the primitive element is computed inside construct domain which uses `ex=True`:
https://github.com/sympy/sympy/blob/96836db06ba6b78103dd4217db53db31502fb2f6/sympy/polys/constructor.py#L105-L128
So actually we already used `ex=True` before the AlgebraicField was even constructed but we didn't pass that information to the AlgebraicField constructor.

In the case in the OP I construct the AlgebraicField and only later ask to convert lots of expressions into it. I think it probably does make sense to use `ex=True` and store the results if a primitive element needs to be constructed in the AlgebraicField constructor.

At the same time though there can be other cases where we have expressions that are not literally part of the primitive element expression but still fit in the domain e.g.:
```python
In [9]: K = QQ.algebraic_field(exp(I*pi/7))

In [10]: K
Out[10]: QQ<exp(I*pi/7)>

In [11]: a = K.from_sympy(cos(2*pi/7)-I*sin(2*pi/7))

In [12]: a
Out[12]: ANP([-1, 0, 0, 0, 0, 0], [1, -1, 1, -1, 1, -1, 1], QQ)

In [13]: K.to_sympy(a)
Out[13]: 
  5⋅ⅈ⋅π
  ─────
    7  
-ℯ  
```
So I think it is still useful to cache the results of `to_number_field` in `K.from_sympy` either way. Actually I just found a bug for this case in this PR...

**Comment by oscarbenjamin** [comment]
It seems that this fails for the PR:
```python
In [3]: K = QQ.algebraic_field(exp(I*pi/7))

In [4]: a = K.from_sympy(cos(2*pi/7)-I*sin(2*pi/7))
---------------------------------------------------------------------------
IsomorphismFailed                         Traceback (most recent call last)
File ~/current/active/sympy/sympy/polys/domains/algebraicfield.py:583, in _make_input_converter.<locals>.isomorphism_converter(a)
    582 try:
--> 583     res = K(to_number_field(a, K.ext).native_coeffs())
    584 except (NotAlgebraic, IsomorphismFailed) as exc:

File ~/current/active/sympy/sympy/polys/numberfields/subfield.py:506, in to_number_field(extension, theta, gen, alias)
    505 else:
--> 506     raise IsomorphismFailed(
    507         "%s is not in a subfield of %s" % (root, theta.root))

IsomorphismFailed: I is not in a subfield of exp(I*pi/7)

The above exception was the direct cause of the following exception:

CoercionFailed                            Traceback (most recent call last)
Cell In[4], line 1
----> 1 a = K.from_sympy(cos(2*pi/7)-I*sin(2*pi/7))

File ~/current/active/sympy/sympy/polys/domains/algebraicfield.py:365, in AlgebraicField.from_sympy(self, a)
    363 if conv is None:
    364     conv = self._input_converter = _make_input_converter(self)
--> 365 return conv(a)

File ~/current/active/sympy/sympy/polys/domains/algebraicfield.py:598, in _make_input_converter.<locals>.converter(a)
    596     result = K.zero
    597     for ai in a.args:
--> 598         result += converter(ai)
    599     return result
    600 elif a.is_Mul:

File ~/current/active/sympy/sympy/polys/domains/algebraicfield.py:603, in _make_input_converter.<locals>.converter(a)
    601     result = K.one
    602     for ai in a.args:
--> 603         result *= converter(ai)
    604     return result
    605 elif a.is_Pow and a.exp.is_Integer:

File ~/current/active/sympy/sympy/polys/domains/algebraicfield.py:612, in _make_input_converter.<locals>.converter(a)
    610 res = cache.get(a)
    611 if res is None:
--> 612     res = cache[a] = isomorphism_converter(a)
    613 return res

File ~/current/active/sympy/sympy/polys/domains/algebraicfield.py:585, in _make_input_converter.<locals>.isomorphism_converter(a)
    583     res = K(to_number_field(a, K.ext).native_coeffs())
    584 except (NotAlgebraic, IsomorphismFailed) as exc:
--> 585     raise CoercionFailed(
    586         "%s is not a valid algebraic number in %s" % (a, K)) from exc
    587 return res

CoercionFailed: I is not a valid algebraic number in QQ<exp(I*pi/7)>
```
On master it gives:
```python
In [9]: K = QQ.algebraic_field(exp(I*pi/7))

In [10]: a = K.from_sympy(cos(2*pi/7)-I*sin(2*pi/7))

In [11]: N(-exp(I*pi/7)**5)
Out[11]: 0.623489801858733 - 0.78183148246803⋅ⅈ

In [12]: N(cos(2*pi/7)-I*sin(2*pi/7))
Out[12]: 0.623489801858733 - 0.78183148246803⋅ⅈ

In [13]: K.from_sympy(I)
...
IsomorphismFailed: I is not in a subfield of exp(I*pi/7)
```
So the bug is trying to convert each part of the expression separately into the number field.

Is it correct that `I` is not in this field?
```python
In [16]: factor(x**2 + 1, extension=exp(I*pi/7))
Out[16]: 
 2    
x  + 1
```
Maybe we can only split out rational factors from a product...

More generally though there is no guarantee that real and imaginary parts of an algebraic number belong to the same field although it works in this case:
```python
In [17]: K.from_sympy(cos(2*pi/7))
Out[17]: ANP([-1/2, 0, 0, 1/2, 0, 0], [1, -1, 1, -1, 1, -1, 1], QQ)

In [18]: K.from_sympy(sin(2*pi/7))
---------------------------------------------------------------------------
IsomorphismFailed 
```
I suppose that when decomposing an Add of Muls we need to make a strictly Q-linear decomposition...

**Comment by oscarbenjamin** [comment]
> I suppose that when decomposing an Add of Muls we need to make a strictly Q-linear decomposition...

Actually there is no guarantee that we can split an Add or a Mul into terms or factors here...

**Comment by oscarbenjamin** [comment]
Marking this as draft. More thought is needed.

Maybe using `ex=True` and handling the primitive element expressions is the right thing and other expressions just have to go through the whole `to_number_field` dance.
